[Cocoa] Audio distortion during media playback when many AudioContexts are created

jernoble opened the pull request:
#### a78127adb38a402b5d0fe6b17367aba32a38eb22
<pre>
[Cocoa] Audio distortion during media playback when many AudioContexts are created
<a href="https://bugs.webkit.org/show_bug.cgi?id=269833">https://bugs.webkit.org/show_bug.cgi?id=269833</a>
<a href="https://rdar.apple.com/122590884">rdar://122590884</a>

Reviewed by Chris Dumez.

In WebKit, each AudioContext created results in an additional thread serving that context&apos;s
AudioDestination. (In WebKitLegacy, each AudioContext will result in an additional
AudioOutputUnit running on a single high-priority audio thread.) When many threads
(or AudioOutputUnits) are created, the overhead alone can cause underruns. And when this
happens on the high-priority audio thread, it affects all audio playback within that
process.

Rather than create new threads or AudioOutputUnits (that are all rendering on the same
thread to the same buffer in the end) for each AudioContext, a shared AudioDestination
can be used for multiple AudioContext&apos;s with the same number of channels and sample rate.
For common scenarios, this means only one high-priority audio thread will be created
and serviced by a single AudioDestination. Specifically for WebKit, it means a single
RemoteAudioDestination/Proxy pair for each WebContent process.

* Source/WebCore/platform/audio/SharedAudioDestination.cpp: Added.
(WebCore::SharedAudioDestinationAdapter::framesPerBuffer const):
(WebCore::SharedAudioDestinationAdapter::sharedMap):
(WebCore::SharedAudioDestinationAdapter::ensureAdapter):
(WebCore::SharedAudioDestinationAdapter::SharedAudioDestinationAdapter):
(WebCore::m_configurationSemaphore):
(WebCore::SharedAudioDestinationAdapter::addRenderer):
(WebCore::SharedAudioDestinationAdapter::removeRenderer):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestination::create):
(WebCore::SharedAudioDestination::SharedAudioDestination):
(WebCore::SharedAudioDestination::~SharedAudioDestination):
(WebCore::SharedAudioDestination::start):
(WebCore::SharedAudioDestination::stop):
(WebCore::SharedAudioDestination::framesPerBuffer const):
(WebCore::SharedAudioDestination::setIsPlaying):
* Source/WebCore/platform/audio/SharedAudioDestination.h: Added.
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestination::create):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):

Canonical link: <a href="https://commits.webkit.org/275262@main">https://commits.webkit.org/275262@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9b5c40c0d747eca45b325022c9ce9a84600d89f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41323 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43887 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37415 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43630 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34174 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41897 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14837 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14992 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36599 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45223 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36918 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40657 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16138 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13244 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39061 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17757 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9270 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+    bool m_started { false };
+
+    // Protected by m_renderLock:
```
That's not how we do these anymore. These things get validated by the compiler now. Please adopt `WTF_GUARDED_BY_LOCK`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static NeverDestroyed<AdapterMap> map;
```
Is this always accessed on the main thread? I assume so since I see no locks.
If so, please use `MainThreadNeverDestroyed`.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+        return Ref { *results->value };
+
+    Ref newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels,  WTFMove(ensureFunction)));
+    sharedMap().set(key, newAdapter);
```
Why `set()` and not `add()`? My suggestion would be to use the more efficient `add()` and add an assertion that the key is not already present in the map.

cdumez replied:
Also, any reason you're not doing this in the constructor like we usually do?

cdumez replied:
I see you adding to the map but never removing, seems fishy.

jernoble replied:
> Why set() and not add()? My suggestion would be to use the more efficient add() and add an assertion that the key is not already present in the map.

Because the key might be in the map, but the value is WeakPtr/null. `add()` would leave the null value in place rather than replace it with a valid value.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        ASSERT(!m_renderers.contains(renderer));
```
How is this safe? what guarantees that `this` is still alive when the lambda runs?

jernoble replied:
Ah looks like I dropped that part during a refactor. There should be a sync dispatch to the configuration queue in the destructor of `SharedAudioDestinationAdapter`.

cdumez replied:
This would work but it seems fragile? Why can't we protect in lambdas?

cdumez replied:
We have well established smart pointer adoption patterns. It would be good to follow them. And if we can't it would be good to understand why so we can tweak our patterns accordingly.

jernoble replied:
Okay, I'll add a WeakPtr/RefPtr lambda check.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+void SharedAudioDestinationAdapter::removeRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        ASSERT(m_renderers.contains(renderer));
```
Same question about lifetime.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+
+void SharedAudioDestinationAdapter::render(AudioBus* sourceBus, AudioBus* destinationBus, size_t numberOfFrames, const AudioIOPosition& outputPosition)
+{
+    if (m_renderLock.tryLock()) {
```
Please use `Locker<Lock>::tryLock()` in new code. No manual locking / unlocking.

jernoble replied:
Oh nice, will do.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,249 @@
+    }
+}
+
+}
```
`// namespace WebCore`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.h:
```diff
@@ -0,0 +1,58 @@
+    Ref<SharedAudioDestinationAdapter> m_outputAdapter;
+};
+
+}
```
`// namespace WebCore`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.h:
```diff
@@ -0,0 +1,58 @@
+    WEBCORE_EXPORT static Ref<SharedAudioDestination> create(AudioIOCallback&, unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&&);
+    WEBCORE_EXPORT virtual ~SharedAudioDestination();
+
+    void ref() const final { return ThreadSafeRefCounted<SharedAudioDestination, WTF::DestructionThread::Main>::ref(); }
```
`<SharedAudioDestination, WTF::DestructionThread::Main>` should not be necessary.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.h:
```diff
@@ -0,0 +1,58 @@
+    WEBCORE_EXPORT virtual ~SharedAudioDestination();
+
+    void ref() const final { return ThreadSafeRefCounted<SharedAudioDestination, WTF::DestructionThread::Main>::ref(); }
+    void deref() const final { return ThreadSafeRefCounted<SharedAudioDestination, WTF::DestructionThread::Main>::deref(); }
```
ditto.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end()) {
+        if (auto existingAdapter = results->value.get())
```
Please use `RefPtr` explicitly instead of `auto` in new code.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+            return existingAdapter.releaseNonNull();
+    }
+
+    Ref<SharedAudioDestinationAdapter> newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels, sampleRate, WTFMove(ensureFunction)));
```
`<SharedAudioDestinationAdapter>` is unnecessary.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+
+    Ref<SharedAudioDestinationAdapter> newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels, sampleRate, WTFMove(ensureFunction)));
+    auto weakAdapter = ThreadSafeWeakPtr<SharedAudioDestinationAdapter> { newAdapter.get() };
+    sharedMap().set(key, WTFMove(weakAdapter));
```
Wouldn't this work?
```
sharedMap().set(key, newAdapter.get());
```
Why do we need to constructor the `ThreadSafeWeakPtr` explicitly. And if we really have to, maybe this would work?
```
ThreadSafeWeakPtr weakAdapter { newAdapter.get() };
```

jernoble replied:
This code was super fiddly, and automatic type deduction was breaking left and right when I wrote it. I originally had this: 

```
sharedMap().set(key, newAdapter.get());
```

And it lead to a compile error deducing the type of the value parameter; hence the rather explicit line 103.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+{
+    auto key = std::make_tuple(m_numberOfOutputChannels, m_sampleRate);
+    sharedMap().remove(key);
+    m_configurationQueue->dispatchSync([] { });
```
Is this really needed now that we capture ThreadSafeWeakPtrs in lambda? I doubt it.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
```
`auto` -> `RefPtr`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
```
You're failing to call the completionHandler here.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
```
Failing to call the completionHandler.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+void SharedAudioDestinationAdapter::removeRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        auto protectedThis = weakThis.get();
```
`RefPtr`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
```
Per the new smart pointer rules, we need a RefPtr *on the stack* to the object you're calling a non-trivial function on. This means, you need a local RefPtr to the WorkQueue before calling `dispatch()` on it. The way we usually do it in new code is adding a new `protectedConfigurationQueue()` member function which returns a `Ref` and call it here: `protectedConfigurationQueue()->dispatch()`.

Same comment applies below.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+{
+    ASSERT(m_configurationQueue->isCurrent());
+    {
+        LockHolder locker { m_renderLock };
```
`LockHolder` is useless in new code, now that we support C++17. We should likely get rid of it. You can just use `Locker` here, which is shorter and equivalent (since `LockHolder` is an alias to `Locker<Lock>` and the template parameter gets deduced with C++17).

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    // Move the previously configured list of renderers to the MainThread for destruction:
+    Vector<RefPtr<SharedAudioDestination>> renderersToDispose;
+    {
+        LockHolder locker { m_renderLock };
```
`Locker`

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    bool firstRenderer = true;
```
Boolean variables need a prefix per WebKit coding style. `isFirstRenderer`?

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+        }
+        // Subsequent renderers should render to the m_workBus, which will
+        // then be summed to the destinationBus.
+        renderer->sharedRender(sourceBus, m_workBus.ptr(), numberOfFrames, outputPosition);
```
You need a Ref to the workBus on the stack before passing it to the function, as per our smart pointer guidelines. Likely want to add a new `protectedWorkBus()` getter which returns a `Ref` and call `protectedWorkBus().ptr()` here.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    bool firstRenderer = true;
+    for (auto& renderer : m_configuredRenderers) {
```
`auto&` -> `RefPtr` for safety, since `m_configuredRenderers` is not on the stack.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+        // Subsequent renderers should render to the m_workBus, which will
+        // then be summed to the destinationBus.
+        renderer->sharedRender(sourceBus, m_workBus.ptr(), numberOfFrames, outputPosition);
+        destinationBus->sumFrom(m_workBus);
```
Ditto, you need a Ref to m_workBus on the stack before passing it.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    setIsPlaying(true);
+    m_outputAdapter->addRenderer(*this, WTFMove(completionHandler));
```
Need a Ref/RefPtr to m_outputAdapter on the stack before calling addRenderer()

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+void SharedAudioDestination::stop(CompletionHandler<void(bool)>&& completionHandler)
+{
+    setIsPlaying(false);
+    m_outputAdapter->removeRenderer(*this, WTFMove(completionHandler));
```
Need a Ref/RefPtr to m_outputAdapter on the stack before calling removeRenderer()

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+
+void SharedAudioDestination::start(Function<void(Function<void()>&&)>&& dispatchToRenderThread, CompletionHandler<void(bool)>&& completionHandler)
+{
+    UNUSED_PARAM(dispatchToRenderThread);
```
Seems unnecessary.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -0,0 +1,315 @@
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
```
`std::tuple key { numberOfOutputChannels, sampleRate };` would likely work nowadays and is slightly shorter.

cdumez reviewed (approved):
LGTM with changes.

jernoble committed:
[Cocoa] Audio distortion during media playback when many AudioContexts are created
https://bugs.webkit.org/show_bug.cgi?id=269833
rdar://122590884

Reviewed by Chris Dumez.

In WebKit, each AudioContext created results in an additional thread serving that context's
AudioDestination. (In WebKitLegacy, each AudioContext will result in an additional
AudioOutputUnit running on a single high-priority audio thread.) When many threads
(or AudioOutputUnits) are created, the overhead alone can cause underruns. And when this
happens on the high-priority audio thread, it affects all audio playback within that
process.

Rather than create new threads or AudioOutputUnits (that are all rendering on the same
thread to the same buffer in the end) for each AudioContext, a shared AudioDestination
can be used for multiple AudioContext's with the same number of channels and sample rate.
For common scenarios, this means only one high-priority audio thread will be created
and serviced by a single AudioDestination. Specifically for WebKit, it means a single
RemoteAudioDestination/Proxy pair for each WebContent process.

* Source/WebCore/platform/audio/SharedAudioDestination.cpp: Added.
(WebCore::SharedAudioDestinationAdapter::framesPerBuffer const):
(WebCore::SharedAudioDestinationAdapter::sharedMap):
(WebCore::SharedAudioDestinationAdapter::ensureAdapter):
(WebCore::SharedAudioDestinationAdapter::SharedAudioDestinationAdapter):
(WebCore::m_configurationSemaphore):
(WebCore::SharedAudioDestinationAdapter::addRenderer):
(WebCore::SharedAudioDestinationAdapter::removeRenderer):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestination::create):
(WebCore::SharedAudioDestination::SharedAudioDestination):
(WebCore::SharedAudioDestination::~SharedAudioDestination):
(WebCore::SharedAudioDestination::start):
(WebCore::SharedAudioDestination::stop):
(WebCore::SharedAudioDestination::framesPerBuffer const):
(WebCore::SharedAudioDestination::setIsPlaying):
* Source/WebCore/platform/audio/SharedAudioDestination.h: Added.
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestination::create):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):

Canonical link: https://commits.webkit.org/275262@main
